File: cs50/week1/problem-set/credit/credit.c
// FILE: Credit
// CREATOR: Tanaka Chitete
// PURPOSE: Validate a credit card using Luhn's algorithm
// CREATION DATE: 24/04/2020
// LATEST MODIFICATION DATE: 25/04/2020

#include <cs50.h>
#include <stdio.h>

int separateDigits(long cardNum, int cardDigs[19]);
int multiplyDigits(int cardDigs[19], int numOfDigs);
int addDigits(int cardDigs[19], int numOfDigs);
int identifyCard(int digOne, int digTwo, int numOfDigs, 
                 int sumLastDig);
void outputCardType(int cardID);

int main(void)
{
    // main variables
    long cardNum;
    int sumTotal, cardDigOne, cardDigTwo, cardID, 
        sumTotalLastDig;

    // Return value assignment variables
    int numOfDigs, sumFromMultiplyDigits, sumFromAddDigits;
    
    cardNum = get_long("Payment card number: ");

    int cardDigs[19];

    // Separates card number into its digits (in reverse order)
    numOfDigs = separateDigits(cardNum, cardDigs);

    if((13 <= numOfDigs) && (numOfDigs <= 16))
    {
        // Sum of multiplying each digit from second-to-last by 2
        sumFromMultiplyDigits = multiplyDigits(cardDigs, numOfDigs);

        // Sum of adding each digit from last
        sumFromAddDigits = addDigits(cardDigs, numOfDigs);

        sumTotal = sumFromMultiplyDigits + sumFromAddDigits;

        // Obtains first and second digits of the credit card for 
        // verification purposes
        cardDigOne = cardDigs[numOfDigs - 1];
        cardDigTwo = cardDigs[numOfDigs - 2];

        // Obtains last digit of sumTotal for verification purposes
        sumTotalLastDig = sumTotal % 10;

        // Obtains ID of card for verification purposes        
        cardID = identifyCard(cardDigOne, cardDigTwo, numOfDigs, 
                              sumTotalLastDig);

        outputCardType(cardID);
    }
    else
    {
        printf("INVALID\n");
    }   
}

// NAME: separateDigits
// IMPORT: cardNum (int), cardDigs (int),  
// EXPORT: numOfDigs (int)
// PURPOSE: Separate digits and store them in array in reverse
// CREATION DATE: 24/04/2020
// LATEST MODIFICATION DATE: 24/04/2020

int separateDigits(long cardNum, int cardDigs[19])
{
    long cardDig;
    int i, numOfDigs;

    i = 0;
    // Loops through each digit of credit card number
    while(cardNum != 0)
    {
        // Obtains current last digit
        cardDig = cardNum % 10;

        cardDigs[i] = cardDig;

        numOfDigs = i + 1;

        // Removes current last digit
        cardNum = cardNum / 10;
        i = i + 1;
    }
    return numOfDigs;
}

// NAME: multiplyDigits
// IMPORT: cardDigs (int), numOfDigs (int)
// EXPORT: sum (int)
// PURPOSE: Calculate sum of every other digit from second-to-last
//          multiplied by two
// CREATION DATE: 24/04/2020
// LATEST MODIFICATION DATE: 25/04/2020

int multiplyDigits(int cardDigs[19], int numOfDigs)
{
    int sum, product, productDigOne, productDigTwo;

    sum = 0;
    for(int i = 1; i < numOfDigs; i += 2)
    {
        product = cardDigs[i] * 2;

        // Splits product into first and second digits if it is too
        // large
        if(product > 9)
        {
            productDigOne = product / 10;
            productDigTwo = product % 10;

            sum = sum + productDigOne + productDigTwo;
        }
        else
        {
            sum = sum + product;
        }
    }
    return sum;
}

// NAME: addDigits
// IMPORT: cardDigs (int), numOfDigs (int)
// EXPORT: sum (int)
// PURPOSE: Calculate sum of every other digit from last
// CREATION DATE: 24/04/2020
// LATEST MODIFICATION DATE: 25/04/2020

int addDigits(int cardDigs[19], int numOfDigs)
{
    int sum;

    sum = 0;
    for(int i = 0; i < numOfDigs; i += 2)
    {
        sum = sum + cardDigs[i];
    }
    return sum;
}

// NAME: identifyCard
// IMPORT: digOne (int), digTwo, numOfDigs (int), sumLastDig (int)
// EXPORT: cardID (int)
// PURPOSE: Card will be identified as either an AMEX, Mastercard
//          or Visa card, otherwise invalid
// CREATION DATE: 25/04/2020
// LATEST MODIFICATION DATE: 25/04/2020

int identifyCard(int digOne, int digTwo, int numOfDigs, 
                 int sumLastDig)
{
    int cardID;

    cardID = 0;
    // Identifies if card is an AMEX card
    if((digOne == 3) &&     
       ((digTwo == 4) || (digTwo == 7)) &&   
       (numOfDigs == 15) &&     
       (sumLastDig == 0))
    {
        cardID = 1;
    }
    // Identifies if card is a Mastercard card
    else if((digOne == 5) && 
            ((1 <= digTwo) && (digTwo <= 5)) && 
            (numOfDigs == 16) && 
            (sumLastDig == 0))
    {
        cardID = 2;
    }
    // Identifies if card is a Visa card
    else if((digOne == 4) && 
            ((numOfDigs == 13) || (numOfDigs == 16)) && 
            (sumLastDig == 0))
    {
        cardID = 3;
    }
    return cardID;
}

// NAME: outputCardType
// IMPORT: cardID (int)
// EXPORT: none
// PURPOSE: Card input will have its outputted
// CREATION DATE: 25/04/2020
// LATEST MODIFICATION DATE: 25/04/2020

void outputCardType(int cardID)
{
    switch(cardID)
    {
        case 1:
            printf("AMEX\n");
            break;
        case 2:
            printf("MASTERCARD\n");
            break;
        case 3:
            printf("VISA\n");
            break;
        default:
            printf("INVALID\n");
            break;
    }
}

File: cs50/week6/problem-set/credit/credit.py
# FILE: Credit
# CREATOR: Tanaka Chitete
# PURPOSE: Validate a credit card implementing Luhn's algorithm
# CREATION: 28/04/2020
# LATEST MODIFICATION: 28/04/2020

from cs50 import get_string

def main():
    cardNum = get_string("Payment card number: ")
    
    cardNum_len = len(cardNum)
    
    # Executes if user input (i.e. 'credit card') is between 13 and 16 digits
    if (13 <= cardNum_len) and (cardNum_len <= 16):
        # Obtains reverse of cardNum
        cardNumRev = cardNum[::-1]

        # Obtains sum of multiplying every other digit from second-to-last by 2
        multDigs_sum = multDigs(cardNumRev, cardNum_len)
        
        # Obtains sum of adding every other digit from last
        addDigs_sum = addDigs(cardNumRev, cardNum_len)
        
        sumTotal = multDigs_sum + addDigs_sum
        
        # Obtains last digit of sumTotal
        sumTotal_digLast = sumTotal % 10
        
        # Executes if last digit of sumTotal is 0
        if sumTotal_digLast == 0:
            # Obtains first and second digits of cardNum
            cardNum_digOne = int(cardNum[0])
            cardNum_digTwo = int(cardNum[1])
            
            # Outputs card type
            outCardType(cardNum_len, cardNum_digOne, cardNum_digTwo)
        # Executes if last digit of sumTotal is not 0
        else:
            print("INVALID")
    # Executes if user input is less than 13 or greater than 16 digits
    else:
        print("INVALID")

# NAME: multDigs
# IMPORT: cardNumRev (str), cardNum_len (int)
# EXPORT: multDigs_sum (int)
# PURPOSE: Return sum of multiplying every other digit from second to-last by 2
# CREATION: 28/04/2020
# LATEST MODIFICATION: 28/04/2020
def multDigs(cardNumRev, cardNum_len):
    multDigs_sum = 0
    for i in range(1, cardNum_len, +2):
        prod = int(cardNumRev[i]) * 2
        
        # Executes if prod is greater than 9
        if prod > 9:
            prodDigOne = prod // 10
            prodDigTwo = prod % 10
            
            multDigs_sum += prodDigOne + prodDigTwo
        # Executes if prod is less than 9
        else:
            multDigs_sum += prod
            
    return multDigs_sum
    
# NAME: addDigs
# IMPORT: cardNumRev (str), cardNum_len (int)
# EXPORT: addDigs_sum (int)
# PURPOSE: Return sum of adding every other digit from last
# CREATION: 28/04/2020
# LATEST MODIFICATION: 28/04/2020
def addDigs(cardNumRev, cardNum_len):
    addDigs_sum = 0
    for i in range(0, cardNum_len, +2):
        addDigs_sum += int(cardNumRev[i])
        
    return addDigs_sum
    
# NAME: outCardType
# IMPORT: cardNum_len (int), cardNum_digOne (int), cardNum_digTwo(int)
# EXPORT: none
# PURPOSE: Output type of card
# CREATION: 28/04/2020
# LATEST MODIFICATION: 28/04/2020
def outCardType(cardNum_len, cardNum_digOne, cardNum_digTwo):
    # Executes if type is American Express
    if cardNum_digOne == 3 and \
       (cardNum_digTwo == 4 or cardNum_digTwo == 7) and \
       cardNum_len == 15:
        print("AMEX")
    # Executes if type is Mastercard
    elif cardNum_digOne == 5 and \
         (1 <= cardNum_digTwo and cardNum_digTwo <= 5) and \
         cardNum_len == 16:
        print("MASTERCARD")
    # Executes if type is Visa
    elif cardNum_digOne == 4 and \
         (cardNum_len == 13 or cardNum_len == 16):
        print("VISA")
    # Executes if type is invalid
    else:
        print("INVALID")
        
        
        
if __name__ == "__main__":
    main()